// @mmisty/cypress-allure-adapter/common/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertToRequestsIncoming = exports.convertToRequestsResponse = void 0;
const convertToRequestsResponse = (data) => {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const url = data.request.url;
    const method = data.request.method;
    const status = (_c = (_b = (_a = data.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.statusCode) !== null && _c !== void 0 ? _c : (_d = data.response) === null || _d === void 0 ? void 0 : _d.statusCode;
    const ended = Date.now(), duration = ended - data.started;
    const res = {
        method: method,
        isFromCypress: false,
        url,
        requestHeaders: data.request.headers,
        requestBody: data.request.body,
        status: status,
        responseHeaders: (_f = (_e = data.response) === null || _e === void 0 ? void 0 : _e.data) === null || _f === void 0 ? void 0 : _f.headers,
        responseBody: (_h = (_g = data.response) === null || _g === void 0 ? void 0 : _g.data) === null || _h === void 0 ? void 0 : _h.body,
        message: `${method} ${status} ${url}`,
        requestStarted: data.started,
        requestEnded: ended,
        duration: duration,
    };
    return res;
};
exports.convertToRequestsResponse = convertToRequestsResponse;
const convertToRequestsIncoming = (data) => {
    const url = data.url;
    const method = data.method;
    const status = undefined;
    const res = {
        method: method,
        isFromCypress: false,
        url,
        requestHeaders: data.headers,
        requestBody: undefined, // todo
        status: status,
        responseHeaders: undefined,
        responseBody: undefined,
        message: `${method} ${status} ${url}`,
        requestStarted: Date.now(),
    };
    return res;
};
exports.convertToRequestsIncoming = convertToRequestsIncoming;
